tests/raftstore_cases/test_region_change_observer.rs

```diff
@@ -0,0 +1,170 @@
+    assert_eq!(event2.0.get_id(), r1);
+    assert_eq!(event2.0.get_peers().len(), 2);
+    assert_ne!(event2.0.get_region_epoch(), event1.0.get_region_epoch());
+
+    // Split
+    cluster.must_put(b"k1", b"v1");
+    cluster.must_put(b"k2", b"v2");
+    cluster.must_put(b"k3", b"v3");
+    cluster.must_split(&event2.0, b"k2");
+    let mut event3 = receiver.recv().unwrap();
+    let mut event4 = receiver.recv().unwrap();
+    // We should receive an `Update` and a `New`. The order of them is not important.
+    if event3.1 != RegionChangeEvent::Update {
+        mem::swap(&mut event3, &mut event4);
+    }
+    // No more events
+    receiver.try_recv().unwrap_err();
+    assert_eq!(event3.1, RegionChangeEvent::Update);
+    assert_eq!(event3.0.get_id(), r1);
+    assert_ne!(event3.0.get_region_epoch(), event2.0.get_region_epoch());
+    let r2 = event4.0.get_id();
+    assert_ne!(r2, r1);
+    assert_eq!(event4.1, RegionChangeEvent::New);
+    if event3.0.get_start_key().is_empty() {
+        assert_eq!(event3.0.get_end_key(), b"k2");
+        assert_eq!(event4.0.get_start_key(), b"k2");
+        assert!(event4.0.get_end_key().is_empty());
+    } else {
+        assert_eq!(event3.0.get_start_key(), b"k2");
+        assert!(event3.0.get_end_key().is_empty());
+        assert!(event4.0.get_start_key().is_empty());
+        assert_eq!(event4.0.get_end_key(), b"k2");
+    }
+
+    // Merge
+    pd_client.must_merge(event3.0.get_id(), event4.0.get_id());
+    // An `Update` produced by PrepareMerge. Ignore it.
+    assert_eq!(receiver.recv().unwrap().1, RegionChangeEvent::Update);
+    let mut event5 = receiver.recv().unwrap();
+    let mut event6 = receiver.recv().unwrap();
+    // We should receive an `Update` and a `Destroy`. The order of them is not important.
+    if event5.1 != RegionChangeEvent::Update {
+        mem::swap(&mut event5, &mut event6);
+    }
+    // No more events
+    receiver.try_recv().unwrap_err();
+    assert_eq!(event5.1, RegionChangeEvent::Update);
+    assert!(event5.0.get_start_key().is_empty());
+    assert!(event5.0.get_end_key().is_empty());
+    assert_eq!(event6.1, RegionChangeEvent::Destroy);
+    if event5.0.get_id() == event3.0.get_id() {
+        assert_eq!(event6.0.get_id(), event4.0.get_id());
+        assert_ne!(event5.0.get_region_epoch(), event3.0.get_region_epoch());
+    } else {
+        assert_eq!(event5.0.get_id(), event4.0.get_id());
+        assert_eq!(event6.0.get_id(), event3.0.get_id());
+        assert_ne!(event5.0.get_region_epoch(), event4.0.get_region_epoch());
+    }
+
+    // Move out from this node
+    // After last time calling "must_add_peer", this region must have two peers
+    assert_eq!(event5.0.get_peers().len(), 2);
+    let r = event5.0.get_id();
+
+    pd_client.must_remove_peer(r, find_peer(&event5.0, 1).unwrap().clone());
+    // TODO: Why is there an `Update`?
+    assert_eq!(receiver.recv().unwrap().1, RegionChangeEvent::Update);
```

Comment:
🤔 Can anyone tell me why removing peer will cause this?

Reply:
Related to merge?

Reply:
I guess it adds a new node first then removes a node.